# IC 4.4 Data Wrangling and Subsetting

# Table of Contents
##### 1. Import libraries
##### 2. Import data
##### 3. Wrangling procedures from CF exercise 4.4
    3.1 3.1 Prep department dictionary
    3.2 Use Data Dictionary to answer questions
##### 4. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.
##### 5. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.
##### 6. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.
##### 7. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.
##### 8. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.
##### 9. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.
##### 10. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?
##### 11. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.
##### 12. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?
##### 13. Export data

## 1. Import libraries

In [3]:
# Imports pandas, numPy, and os libraries

import pandas as pd
import numpy as np
import os

## 2. Import data

In [5]:
path = r'C:\Users\ashle\Documents\07-2024 Instacart Basket Analysis' # Assigns the master file path to variable 'path'
vars_list = ['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day', 'days_since_prior_order'] # Assigns the necessary columns to vars_list

# Imports orders.csv using the os.path.join() function and vars_list, omitting the "eval_set" column.
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False, usecols = vars_list)

# Imports products.csv using the 'path' variable. 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

#Imports departments.csv using the 'path' variable.
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

## 3. Wrangling procedures from CF exercise 4.4

In [13]:
#look for any columns that need to be renamed
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [11]:
#checking frequency on 'days_since_prior_order' column regarding NaN values found. 
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [15]:
# There were 206,209 Nan values discovered.

### 3.1 Prep department dictionary

In [38]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

Note the wide format of the department dataframe. To handle this, we will transpose it

In [20]:
#Transposing df_dep
df_dep_t = df_dep.T

In [42]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [22]:
#add an index column
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [28]:
# Take the first row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [48]:
new_header

0    department
Name: department_id, dtype: object

In [24]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [52]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [30]:
#set the header row as the df header
df_dep_t_new.columns = new_header

In [56]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [32]:
# Turn the df_dep_t_new dataframe into a dictionary
data_dict = df_dep_t_new.to_dict('index')

In [64]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

### 3.2 Use Data Dictionary to answer questions

In [66]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [41]:
#identify what department_id '19' represents
print(data_dict.get('19'))

{'department': 'snacks'}


In [43]:
#creates a subset df_snacks containing all products under the 'snack' department
df_snacks = df_prods[df_prods['department_id']==19]

In [45]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## 4. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

Begin by reviewing data types for each variable. 
Order_id and user_id would not negatively impact analysis if changed to a non-numerical format. 
For this exercise I will change order_id to a string. 

In [104]:
df_ords.dtypes #view data types for each variable

order_id                   object
user_id                     int64
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [107]:
#changes data type of 'order_id' in df_ords to a string and assigns it to the 'order_id' column in df_ords
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [101]:
#checks to ensure data type was successfully changed
df_ords.dtypes

order_id                   object
user_id                     int64
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

The data type for order_id is now object (which is pandas way of representing a string).

## 5. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

Begin by viewing column names. Change 'order_dow' to something more intuitive like 'order_day_of_week'. Save this column name as a "new" column. 

In [118]:
df_ords.columns #view column names

Index(['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day',
       'days_since_prior_order'],
      dtype='object')

In [120]:
#renames 'order_dow' to 'order_day_of_week' column in the df_ords dataframe. 
#this does not overwrite the dataframe because we did not reassign it or use the 'inplace=True' argument.
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'})

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0        2539329        1             1                  2                  8   
1        2398795        1             2                  3                  7   
2         473747        1             3                  3                 12   
3        2254736        1             4                  4                  7   
4         431534        1             5                  4                 15   
...          ...      ...           ...                ...                ...   
3421078  2266710   206209            10                  5                 18   
3421079  1854736   206209            11                  4                 10   
3421080   626363   206209            12                  1                 12   
3421081  2977660   206209            13                  1                 12   
3421082   272231   206209            14                  6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

## 6. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [136]:
df_ords['order_hour_of_day'].value_counts() #counts the number of occurances for each value in the 'order_hour_of_day' column in df_ords.

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

The busiest hour for placing orders is 10 AM.

## 7. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

The data dictionary for departments was created in the Data Wrangling section. We will reference it here. 

In [142]:
print(data_dict.get('4')) #returns value that corresponds with department_id '4'

{'department': 'produce'}


The meaning behind the value of 4 in the 'department_id' column in the df_prods dataframe is produce.

## 8. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [161]:
#filters the dataframe to find the department_id for "breakfast"
df_dep_t_new[df_dep_t_new['department'] == 'breakfast']

department_id department
14             breakfast

breakfast = department_id 14

In [159]:
#creates subset df_breakfast where 'department_id' is equal to 14 in the df_prods dataframe
df_breakfast = df_prods[df_prods['department_id']==14]

In [167]:
df_breakfast #pulls subset df_breakfast dataframe to ensure it was created properly. 

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

## 9. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

First step is to identify department ids for alcohol, deli, beverages, and meat/seafood.

In [179]:
#filters df_dep_t_new dataframe to find department_ids for alcohol, deli, beverages, and meat/seafood

df_dep_t_new[df_dep_t_new['department'].isin(['alcohol', 'deli', 'beverages', 'meat seafood'])]

department_id    department
5                   alcohol
7                 beverages
12             meat seafood
20                     deli

In [187]:
#use isin() function to create subset for multiple departments. 

df_dinnerparty = df_prods[df_prods['department_id'].isin([5, 7, 12, 20])]

In [189]:
df_dinnerparty #pulls subset df_breakfast dataframe to ensure it was created properly. 

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

## 10. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [200]:
df_dinnerparty.shape #use shape to show the number of rows and columns. 

(7650, 5)

There are 7650 rows in the last dataframe I created 'df_dinnerparty'. It also displayed at the bottom of the command df_dinnerparty from Step 7. 

## 11. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [226]:
user_id1 = df_ords[df_ords['user_id']==1] #filters df_ords dataframe for user_id that is equal to 1. 
user_id1

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   
5   3367565        1             6          2                  7   
6    550135        1             7          1                  9   
7   3108588        1             8          1                 14   
8   2295261        1             9          1                 16   
9   2550362        1            10          4                  8   
10  1187899        1            11          4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

## 12. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [230]:
#you can manually calculate these since the dataframe is so small, however we could use user_id1.describe()
user_id1.describe()

user_id  order_number  order_dow  order_hour_of_day  \
count     11.0     11.000000  11.000000          11.000000   
mean       1.0      6.000000   2.636364          10.090909   
std        0.0      3.316625   1.286291           3.477198   
min        1.0      1.000000   1.000000           7.000000   
25%        1.0      3.500000   1.500000           7.500000   
50%        1.0      6.000000   3.000000           8.000000   
75%        1.0      8.500000   4.000000          13.000000   
max        1.0     11.000000   4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

User_id 1 has made 11 orders in total.

Average days between orders: 19

Median days between orders: 19.5

Most common day of week orders were made: 4

Most common hour of day orders were made: 7 AM

Longest number of days between orders: 30 days

Shortest number of days between orders: 0 days

## 13. Export Data

In [47]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [49]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))